pex/jobs.py

```diff
@@ -0,0 +1,207 @@
+  :type interpreter: :class:`PythonInterpreter`
+  :param expose: The names of any vendored distributions to expose to the spawned python process.
+  :type expose: list of str
+  :param subprocess_kwargs: Any additional :class:`subprocess.Popen` kwargs to pass through.
+  :returns: A job handle to the spawned python process.
+  :rtype: :class:`Job`
+  """
+  if expose:
+    subprocess_env = (env or os.environ).copy()
+    subprocess_env['__PEX_UNVENDORED__'] = '1'
+
+    pythonpath = third_party.expose(expose)
+  else:
+    subprocess_env = env
+    pythonpath = None
+
+  interpreter = interpreter or PythonInterpreter.get()
+  cmd, process = interpreter.open_process(
+    args=args,
+    pythonpath=pythonpath,
+    env=subprocess_env,
+    **subprocess_kwargs
+  )
+  return Job(command=cmd, process=process)
+
+
+class SpawnedJob(namedtuple('SpawnedJob', ['job', 'result_func'])):
+  """A handle to a spawned :class:`Job` and its associated result."""
+
+  @classmethod
+  def wait(cls, job, result):
+    def wait_result_func():
+      job.wait()
+      return result
+
+    return cls(job=job, result_func=wait_result_func)
+
+  @classmethod
+  def stdout(cls, job, result_func, input=None):
+    def stdout_result_func():
+      stdout, _ = job.communicate(input=input)
+      return result_func(stdout)
+
+    return cls(job=job, result_func=stdout_result_func)
+
+  def await_result(self):
+    """Waits for the spawned job to complete and returns its result."""
+    return self.result_func()
+
+  def kill(self):
+    """Terminates the spawned job if it's not already complete."""
+    self.job.kill()
+
+
+_CPU_COUNT = cpu_count()
+_ABSOLUTE_MAX_JOBS = _CPU_COUNT * 2
+
+
+DEFAULT_MAX_JOBS = _CPU_COUNT
+"""The default maximum number of parallel jobs PEX should use."""
+
+
+def _sanitize_max_jobs(max_jobs=None):
+  assert max_jobs is None or isinstance(max_jobs, int)
+  if max_jobs is None or max_jobs <= 0:
+    return DEFAULT_MAX_JOBS
+  else:
+    return min(max_jobs, _ABSOLUTE_MAX_JOBS)
+
+
+def execute_parallel(max_jobs, inputs, spawn_func, raise_type):
+  """Execute jobs for the given inputs in parallel.
+
+  :param int max_jobs: The maximum number of parallel jobs to spawn.
+  :param inputs: An iterable of the data to parallelize over `spawn_func`.
+  :param spawn_func: A function taking a single input and returning a :class:`SpawnedJob`.
+  :param raise_type: A type that takes a single string argument and will be used to construct a
+                     raiseable value when any of the spawned jobs errors.
+  :returns: An iterator over the spawned job results as they come in.
+  :raises: A `raise_type` exception if any individual job errors.
+  """
+  size = _sanitize_max_jobs(max_jobs)
+  TRACER.log('Spawning a maximum of {} parallel jobs to process:\n  {}'
+             .format(size, '\n  '.join(map(str, inputs))),
+             V=9)
+
+  stop = Event()
+  job_slots = BoundedSemaphore(value=size)
+  spawned_job_queue = Queue()
```

Comment:
As discussed, this should actually be `Queue[Union[SpawnedJob, Exception, object]]`

Reply:
I went with `Queue[Union[SpawnedJob, Exception, Literal[done_sentinel]]]` even though `Literal` is experimental / will likely never support values like `object()` since `object` is clear as mud if you pause on it for a second.

Reply:
Ah much better. Thank you!

I'm not concerned about MyPy recognizing this as valid. We're not actually using MyPy. Rather, the reason I pitched a type comment is that it's now a standardized way in Python to express type information.